[rocblas] build fails with GCC 11

GCC 11 seems to have changed their header files, I get this build error:

```
In file included from /build/rocblas/src/build/virtualenv/lib/python3.9/site-packages/Tensile/Source/lib/source/TensorDescriptor.cpp:35:
In file included from /build/rocblas/src/build/virtualenv/lib/python3.9/site-packages/Tensile/Source/lib/include/Tensile/TensorDescriptor.hpp:38:
In file included from /build/rocblas/src/build/virtualenv/lib/python3.9/site-packages/Tensile/Source/lib/include/Tensile/DataTypes.hpp:29:
In file included from /opt/rocm/llvm/lib/clang/12.0.0/include/cuda_wrappers/complex:75:
/usr/lib64/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/complex:703:7: error: __host__ __device__ function '__failed_assertion' cannot overload __host__ function '__failed_assertion'
      __glibcxx_assert( __rho >= 0 );
      ^
/usr/lib64/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/x86_64-pc-linux-gnu/bits/c++config.h:754:5: note: expanded from macro '__glibcxx_assert'
    __glibcxx_assert_1(_Condition)        \
    ^
/usr/lib64/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/x86_64-pc-linux-gnu/bits/c++config.h:743:13: note: expanded from macro '__glibcxx_assert_1'
       void __failed_assertion();       \
            ^
/usr/lib64/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/bits/std_mutex.h:142:7: note: previous declaration is here
      __glibcxx_assert(__e != EBUSY); // threads are still blocked
      ^
/usr/lib64/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/x86_64-pc-linux-gnu/bits/c++config.h:754:5: note: expanded from macro '__glibcxx_assert'
    __glibcxx_assert_1(_Condition)        \
    ^
/usr/lib64/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/x86_64-pc-linux-gnu/bits/c++config.h:743:13: note: expanded from macro '__glibcxx_assert_1'
       void __failed_assertion();       \
            ^
1 error generated when compiling for host.
make[2]: *** [Tensile/lib/CMakeFiles/TensileHost.dir/build.make:216: Tensile/lib/CMakeFiles/TensileHost.dir/source/TensorDescriptor.cpp.o] Error 1
make[2]: Leaving directory '/build/rocblas/src/build'
make[1]: *** [CMakeFiles/Makefile2:205: Tensile/lib/CMakeFiles/TensileHost.dir/all] Error 2
make[1]: Leaving directory '/build/rocblas/src/build'
make: *** [Makefile:156: all] Error 2
make: Leaving directory '/build/rocblas/src/build'
```

--- a/llvm-amdgpu/.SRCINFO
+++ b/llvm-amdgpu/.SRCINFO
@@ -1,7 +1,7 @@
 pkgbase = llvm-amdgpu
 	pkgdesc = Radeon Open Compute - LLVM toolchain (llvm, clang, lld)
 	pkgver = 4.2.0
-	pkgrel = 1
+	pkgrel = 2
 	url = https://github.com/RadeonOpenCompute/llvm-project
 	arch = x86_64
 	license = custom:Apache 2.0 with LLVM Exception
@@ -11,6 +11,8 @@ pkgbase = llvm-amdgpu
 	depends = z3
 	options = staticlibs
 	source = llvm-amdgpu-4.2.0.tar.gz::https://github.com/RadeonOpenCompute/llvm-project/archive/rocm-4.2.0.tar.gz
+	source = libstdc++-11.1.0-workaround.patch::https://reviews.llvm.org/file/data/77rbtfy27xsj2sgfygq5/PHID-FILE-7vya6mqfpldq7sohw56a/file
 	sha256sums = 751eca1d18595b565cfafa01c3cb43efb9107874865a60c80d6760ba83edb661
+	sha256sums = c874456a8a616735bf524772f7ca229d8e59d9775b2b5f19d734df1afda57d11
 
 pkgname = llvm-amdgpu

--- a/llvm-amdgpu/PKGBUILD
+++ b/llvm-amdgpu/PKGBUILD
@@ -4,16 +4,25 @@
 pkgname=llvm-amdgpu
 pkgdesc='Radeon Open Compute - LLVM toolchain (llvm, clang, lld)'
 pkgver=4.2.0
-pkgrel=1
+pkgrel=2
 arch=('x86_64')
 url='https://github.com/RadeonOpenCompute/llvm-project'
 license=('custom:Apache 2.0 with LLVM Exception')
 depends=(z3)
 makedepends=(cmake python ninja)
-source=("${pkgname}-${pkgver}.tar.gz::$url/archive/rocm-$pkgver.tar.gz")
-sha256sums=('751eca1d18595b565cfafa01c3cb43efb9107874865a60c80d6760ba83edb661')
+source=("${pkgname}-${pkgver}.tar.gz::$url/archive/rocm-$pkgver.tar.gz"
+        "libstdc++-11.1.0-workaround.patch::https://reviews.llvm.org/file/data/77rbtfy27xsj2sgfygq5/PHID-FILE-7vya6mqfpldq7sohw56a/file")
+sha256sums=('751eca1d18595b565cfafa01c3cb43efb9107874865a60c80d6760ba83edb661'
+            'c874456a8a616735bf524772f7ca229d8e59d9775b2b5f19d734df1afda57d11')
 options=(staticlibs)
 _dirname="$(basename "$url")-$(basename "${source[0]}" .tar.gz)"
+
+prepare() {
+    cd "$_dirname"
+
+    # fix compatibility with GCC 11: https://github.com/ROCmSoftwarePlatform/rocBLAS/issues/1191#issuecomment-851634017
+    patch -Np1 < "$srcdir/libstdc++-11.1.0-workaround.patch"
+}
 
 build() {
     cmake -GNinja \

